src/testing/vortex/faulty_network.zig

```diff
@@ -0,0 +1,583 @@
+//! The `Network` is a set of proxies used to inject network faults in a Vortex test cluster. We
+//! create one `Proxy` per replica in the test cluster. Each proxy has a set of available
+//! `Connection`s, which model the communication through the proxy (replica-to-replica or
+//! client-to-replica). Each `Connection` has two pipes, connecting the peers' inputs and outputs
+//! (recv and send).
+//!
+//! The _mappings_ are pairs of addresses:
+//!
+//! * _origin_: the address on which the proxy listens for connections from _origin_ peers
+//! * _remote_: the address the proxy connects to, to communicate with the _remote_ peers
+//!
+//! A pipe runs, alternating recv and send, until it sees that the connection is no longer proxying
+//! or if there's an error or EOF, in which case it also closes the parent connection.
+//!
+//! When the `Faults` struct is populated with non-null configurations, the pipes inject
+//! corresponding faults according to the probabilities.
+//!
+//! NOTE: The pipe is not yet message-aware (and perhaps shouldn't be?), which means that we deal
+//! with whatever chunks of bytes we receieve immediately, instead of collecting a buffer with a
+//! full message before piping it through.
```

Comment:
As a note, the sort of faulting we do here would appear very differently to what underlying network packet loss / corruption / latency would do to TCP, since we're simulating those things on top of TCP rather than below TCP (and seeing how the TCP state machine reacts too).

I don't think it's a problem, or that there's an easy solution, but something interesting to think about!

Reply:
Hmm - not without getting crazy, I think :smile: perhaps with a custom TUN/TAP pair or something similar.

Reply:
>I don't think it's a problem, or that there's an easy solution, but something interesting to think about!

I think we can just use a RAW socket to mess up at the IP layer? I think it'll require the faulty_network to understand some parts of the TCP state machine, but it shouldn't be anything crazy in terms of the kernel APIs used?